MapboxDirections/MBDirectionOptions.swift

```diff
@@ -0,0 +1,454 @@
+        coder.encode(includesSteps, forKey: "includesSteps")
+        coder.encode(shapeFormat.description, forKey: "shapeFormat")
+        coder.encode(routeShapeResolution.description, forKey: "routeShapeResolution")
+        coder.encode(attributeOptions.description, forKey: "attributeOptions")
+        coder.encode(locale, forKey: "locale")
+        coder.encode(includesSpokenInstructions, forKey: "includesSpokenInstructions")
+        coder.encode(distanceMeasurementSystem.description, forKey: "distanceMeasurementSystem")
+        coder.encode(includesVisualInstructions, forKey: "includesVisualInstructions")
+    }
+    
+    public required init?(coder decoder: NSCoder) {
+        guard let waypoints = decoder.decodeObject(of: [NSArray.self, Waypoint.self], forKey: "waypoints") as? [Waypoint] else {
+            return nil
+        }
+        self.waypoints = waypoints
+        
+        guard let profileIdentifier = decoder.decodeObject(of: NSString.self, forKey: "profileIdentifier") as String? else {
+            return nil
+        }
+        self.profileIdentifier = MBDirectionsProfileIdentifier(rawValue: profileIdentifier)
+        
+        includesSteps = decoder.decodeBool(forKey: "includesSteps")
+        
+        guard let shapeFormat = RouteShapeFormat(description: decoder.decodeObject(of: NSString.self, forKey: "shapeFormat") as String? ?? "") else {
+            return nil
+        }
+        self.shapeFormat = shapeFormat
+        
+        guard let routeShapeResolution = RouteShapeResolution(description: decoder.decodeObject(of: NSString.self, forKey: "routeShapeResolution") as String? ?? "") else {
+            return nil
+        }
+        self.routeShapeResolution = routeShapeResolution
+        
+        guard let descriptions = decoder.decodeObject(of: NSString.self, forKey: "attributeOptions") as String?,
+            let attributeOptions = AttributeOptions(descriptions: descriptions.components(separatedBy: ",")) else {
+                return nil
+        }
+        self.attributeOptions = attributeOptions
+        
+        
+        if let locale = decoder.decodeObject(of: NSLocale.self, forKey: "locale") as Locale? {
+            self.locale = locale
+        }
+        
+        includesSpokenInstructions = decoder.decodeBool(forKey: "includesSpokenInstructions")
+        
+        if let distanceMeasurementSystem = MeasurementSystem(description: decoder.decodeObject(of: NSString.self, forKey: "distanceMeasurementSystem") as String? ?? "") {
+            self.distanceMeasurementSystem = distanceMeasurementSystem
+        }
+        
+        includesVisualInstructions = decoder.decodeBool(forKey: "includesVisualInstructions")
+    }
+    
+    /**
+     An array of directions query strings to include in the request URL.
+     */
+    internal var queries: [String] {
+        return waypoints.map { "\($0.coordinate.longitude),\($0.coordinate.latitude)" }
+    }
+    
+    internal var path: String { return "" }
```

Comment:
Unsure what to do here. We need this prop on DirectionOptions but _always_ overiden by the class which subclasses it.

Reply:
Yeah, there’s no way in Objective-C to make an abstract class’s instance method purely virtual, as in C++, since “abstract class” is really just a convention rather than a compiler feature. In this situation, you could assert that this abstract method should never be called directly, then return an empty string.

On the other hand, this is a good case for using a protocol instead of an abstract class. The only difficulty with using a protocol is that we can’t provide a default implementation in an Objective-C-compatible protocol, so there would be a lot of code duplication. We’d have to find alternative ways to share code between RouteOptions and MatchOptions.